Fixed #29817 -- Deprecated settings.FILE_CHARSET.

jdufresne opened the pull request:
https://code.djangoproject.com/ticket/29817

carltongibson commented:
Closing as per ticket. (Again, happy to reopen if the [django-developers discussion](https://groups.google.com/d/topic/django-developers/7bZbYVV6hg4/discussion) evolves suitably.)

jdufresne commented:
Ticket was reopened, reopening PR too.

timgraham commented:
Could you remove mentions of this setting in the documentation as much as possible (for example, docs/ref/unicode.txt should mention TEMPLATES instead)?

carltongibson commented:
I had one thought, do we need to state anywhere that templates etc **must** now be UTF-8?

jdufresne commented:
@timgraham I have adjusted the docs to try to be more accurate and remove references to `FILE_CHARSET`. Let me know if you'd like any additional edits.

@carltongibson Do the changes to `unicode.txt` address your feedback? Or does the [existing description of `file_charset`](https://github.com/django/django/blob/1b49b792e9f0e6a5ae22f49f75586847e7e183bf/docs/topics/templates.txt#L352-L354)? Please let me know if you had something else in mind.

carltongibson commented:
Hey @jdufresne: I think the changes to `file_charset` are fine. Thanks! (The user I have in mind **is** using UTF-8. The ref in the TEMPLATES config is thus sufficient.) 👍

timgraham commented:
I guess [whitenoise](https://github.com/evansd/whitenoise/blame/fa330ea1193da72dc9d65f73223d4b30a4f8abb3/whitenoise/middleware.py#L73) should switch to using 'utf-8' as a default for [WHITENOISE_CHARSET](http://whitenoise.evans.io/en/stable/django.html#WHITENOISE_CHARSET) rather than`settings.FILE_CHARSET`. Agreed?

I guess we can't raise a deprecation warning for `FILE_CHARSET` accesses without capturing the warning in Django's usage. I don't remember if I intentionally skipped that for the `DEFAULT_CONTENT_TYPE` deprecation. Not sure if it's worth it.

jdufresne commented:
> I guess whitenoise should switch to using 'utf-8' as a default for WHITENOISE_CHARSET rather thansettings.FILE_CHARSET. Agreed?

That sounds correct to me, but I'm not very familiar with whitenoise or its code. Your proposal there sounds very sensible.

jdufresne commented:
@timgraham I added a new commit that warns on `FILE_CHARSET` access (we could use the same strategy  for `DEFAULT_CONTENT_TYPE` too), but avoids false positives originating from Django. Let me know what you think. If you think it is a good idea, we can squash it, otherwise drop it.

timgraham commented:
Yes, I think it would be useful as I found some uses of `FILE_CHARSET` in [other code](https://github.com/search?q=%22settings.FILE_CHARSET%22+language%3APython+language%3APython&type=Code). If you could use the same technique for `DEFAULT_CONTENT_TYPE` (in a separate PR of course), that would [be useful](https://github.com/search?q=%22settings.DEFAULT_CONTENT_TYPE%22+-filename%3Atest_default_content_type.py+language%3APython+language%3APython&type=Code).

timgraham commented on django/conf/__init__.py:
```diff
@@ -93,6 +97,22 @@ def configured(self):
+        filename, _line_number, _function_name, _text = stack[-2]
+        if not filename.startswith(os.path.dirname(django.__file__)):
+            warnings.warn(
+                'The FILE_CHARSET setting is deprecated. Starting with Django '
```
Maybe this could be a module constant so as not to repeat it 3 times.

timgraham commented on django/conf/__init__.py:
```diff
@@ -93,6 +97,22 @@ def configured(self):
+    @property
+    def FILE_CHARSET(self):
+        stack = traceback.extract_stack()
+        # Stack index:
```
```
# Show a warning if the setting is used outside of Django.
# Stack index: -1 this line, -2 the caller.
```

timgraham commented on tests/settings_tests/test_deprecation.py:
```diff
@@ -0,0 +1,43 @@
+    def test_access_warning(self):
+        with self.assertRaisesMessage(RemovedInDjango31Warning, self.msg):
+            settings.FILE_CHARSET
+        # Works a second time.
```
Curious about the reason why it may not work a second time (and it's unclear to me that it needs to work a second time, maybe one warning is enough?).

jdufresne commented:
Changes:

- Squashed commits
- Added module constant `FILE_CHARSET_DEPRECATED_MESSAGE` to avoid repeating the string
- Improved explanatory comment as suggested

Thanks

jdufresne commented:
> If you could use the same technique for DEFAULT_CONTENT_TYPE (in a separate PR of course), that would be useful.

Done in #10513
